### DISCUSSION TITLE
Substring function?

### DISCUSSION THREAD

**Discussion by vaadin-miki** [community/community#26625] [Actions]
<p>Hello there,</p>
<p>Assume an action that triggers when a PR is made to a branch “ticket-123”, where “ticket-” is fixed. I can easily make a conditional step, using “if: startsWith()”.</p>
<p>Question: I need to pass “123” as a parameter to an action:</p>
<pre><code>- uses: an/action
  with:
    parameter-name: ???
</code></pre>
<p>How do I get “123” from “ticket-123”. I cannot seem to find in the docs anything about substring being supported, so there should be some other way.</p>
<p>Any help will be greatly appreciated. Thanks!</p>

**Comment by jclem** [comment, marked as answer]
<p>Unfortunately, there’s currently no built-in substring function. As a workaround, you can use another action and set the substring as an output variable:</p>
<pre><code>- id: get-id
  run: |
    id=$(echo ${{github.event.client_payload.value}} | cut -d- -f2)
    echo "::set-output name=id::$id"
- run: echo "${{steps.get-id.outputs.id}}"
</code></pre>
<p>In this example, I’ve added a step with the ID <em>get-id</em>. That step splits the value that I want to split using cut and then <a href="https://help.github.com/en/actions/reference/workflow-commands-for-github-actions#setting-an-output-parameter" rel="nofollow noopener">sets it as an output parameter</a>, which can be read and used by subsequent actions.</p>
<p>If you’re not familiar with cut, it’s a shell command that splits a string based on a delimiter (-d) and outputs a specific field (-f). You could use any <a href="https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#using-a-specific-shell" rel="nofollow noopener">shell</a> we support, or you could even easily use JavaScript with the <a href="https://github.com/actions/github-script" rel="nofollow noopener">actions/github-script </a>action:</p>
<pre><code>- id: get-id
  uses: actions/github-script@0.9.0
  with:
    script: |
      const id = context.payload.client_payload.value.split('-')[1]
      core.setOutput('id', id)
- run: echo "${{steps.get-id.outputs.id}}"
</code></pre>

**Reply by ssbarnea** [reply]
Hilarious how crippled the github expressions are. Not even ability to cut a string to size.

**Reply by jclem** [reply]
I don't quite understand the "hilarious" part of your comment. I can definitely see a substring function being useful in a good number of cases, though! I just don't see how the absence of it ranks as hilarious. Thankfully in the meantime, there are (admittedly somewhat cumbersome) workarounds.

**Reply by ulzha** [reply]
@ssbarnea TBH there's a point to be made that build pipelines deserve to be simple and declarative. And there's a point to be made about how (re)inventing additional expression languages hardly adds value to the software development community - we can use existing battle proven languages instead - bash, python, what have you. Actions can work with outputs/env vars from scripts executed in prior steps. Compose tools that we have, instead of expecting each tool to turn into a (new and slightly divergent brand of) Swiss army knife.

**Reply by cberg-aot** [reply]
@ulzha Its a templating engine, just like everything else!  

As an example, Helm will support a combination of Go and Sprig functions out of the box. 
![image](https://user-images.githubusercontent.com/93226309/216478467-a99d5200-c258-45fc-ac64-6137a2bb7c9b.png)

Last I checked @ssbarnea had a very normal use for traditional pipelines!  Missing inline string manipulation is indeed hilarious

**Reply by laeubi** [reply]
I just wanted to leave a note that set-output is now deprecated:
https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/

beside that, even though its not that comfortable the proposed solution works and opens the area for a wide range of use-cases!

**Reply by TWiStErRob** [reply]
@laeubi only the `::set-output` workflow command is deprecated; the `core.setOutput` function and the `>> $GITHUB_OUTPUT` environment file syntax are still very much alive and not going anywhere.

So to rephrase the shell workaround following the deprecation:

```yml
- id: get-id
  run: |
    id=$(echo ${{github.event.client_payload.value}} | cut -d- -f2)
    echo "id=$id" >> $GITHUB_OUTPUT
- run: echo "${{steps.get-id.outputs.id}}"
```
and to use the `core.setOutput` variant, make sure you're running `actions/github-script` on a compatible version where `@action/core v1.10.0` or newer is used, so minimum: `2.1.0`, `3.2.0`, `4.2.0`, `5.2.0`, `6.3.2` (yeah, they backported it on almost all github-script versions 👏)

**Reply by hfhbd** [reply]
Are there any plans to provide such substring or even regex replace function in the future? This would allow you to use the substring function in all expression locations.
One common use-case is to get the version (1.0.0) of a git tag (v1.0.0) which starts with v to allow protected tags v*. At the moment you have to use a step:
`run: long="${{ github.ref }}"; version=${long#"refs/tags/v"}; echo "version=${version}" >> $GITHUB_ENV`
But this only sets a job local environment variable and you can't conditionally start a job (eg different jobs for 1.0 or 2.0).

**Reply by TWiStErRob** [reply]
@hfhbd As workaround's workaround, you can have a separate job with this step, expose the step output as a job output, and use `needs` to `if:` on multiple jobs; resulting in the ability to only execute one of the jobs.

I agree this is waaaaay more than it should be necessary to achieve this use case! 😔

**Reply by hfhbd** [reply]
Yes, I am aware of this workaround I current use. But there is always hope for a better solution using a built-in function.

**Reply by zackw** [reply]
The workaround of setting variables from steps doesn't help me because it cannot be used to set values that have to be available _before the first step executes_, such as the `name` of the entire job.  Concrete example:

```
jobs:
  conda-unit:
    name: |
      ${{ format('conda python: {0} os: {1} env: {2}',
                 matrix.python-version, matrix.os, matrix.environment-file) }}
    strategy:
      fail-fast: false
      matrix:
        os: [ubuntu-latest, macos-latest, windows-latest]
        environment-file: ["environment.yml", "minimal_environment.yml"]
        python-version: ["3.9", "3.10", "3.11", "3.12", "3.13.0-rc.1"]
```

This works but the job names it generates are too long to fit in the fairly narrow job list column, and getting rid of the redundant "-latest" and "environment.yml" would really help.

**Reply by owattley-rotageek** [reply]
@zackw gets it. Expressions really need to be expanded so we can do "advanced" (😂) things like get dates, substring, replace, etc, etc. I use top level workflow `run-name:` to offer up build names to my engineers so they can copy/paste from the workflow runs overview page without having to drill into runs and it's SO limiting what you are able to place in the `run-name:` it makes me tear my hair out.

**Reply by antpaw** [reply]
updated version:
```
      - name: Shorten commit SHA
        uses: actions/github-script@ed597411d8f924073f98dfc5c65a23a2325f34cd
        id: set-short-sha
        with:
          script: return context.sha.substring(0, 8)
          result-encoding: string
```

use the result like this `${{ steps.set-short-sha.outputs.result }}`

**Comment by vaadin-miki** [comment]
<p><span class="mention">@jclemthank</span> you very much :slight_smile: Using outputs from one step as inputs to another will work perfectly fine as a workaround!</p>

**Comment by marchenko1985** [comment]
<p>Thank you, <code>github-script</code> action is awesome, if you take a closer look at what it can do, you can do some fun things like:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/a/2/a2f4e8a630dc2093eeb0370eb100abbda0512825.png" data-download-href="/uploads/short-url/nfA41lvsAgnyZ0cVhREYw5vq0DP.png?dl=1" title="image" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/88868/181098748-89a80839-4ecf-4346-a926-9da88680626e.png" alt="image" data-base62-sha1="nfA41lvsAgnyZ0cVhREYw5vq0DP" width="690" height="400" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/a/2/a2f4e8a630dc2093eeb0370eb100abbda0512825_2_690x400.png, https://aws1.discourse-cdn.com/github/original/3X/a/2/a2f4e8a630dc2093eeb0370eb100abbda0512825.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/a/2/a2f4e8a630dc2093eeb0370eb100abbda0512825.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/a/2/a2f4e8a630dc2093eeb0370eb100abbda0512825_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">956×555 74.4 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>but, what is even better, you can store this script in file in your repository and test it locally by just running it with required environment variables being defined with <code>INTPUT_</code> prefix</p>

**Comment by jdrphillips** [comment]
Now you can have named runs based on github context, a built-in substring function would be very useful. For example using `{{ github.sha }}` in the name makes it unreadable as the sha is so long. It would be great to be able to cut that to 7 characters inside the custom run name

**Reply by tgillitzer** [reply]
This is exactly what I was looking to do when I found this.

Ansible uses Jinja to handle its "expressions", which basically uses the exact same syntax.   Would be cool if we could just use Jinja.

**Reply by bayocr** [reply]
I was looking for the same for a while and I didn't find a way.

**Reply by salvianreynaldi** [reply]
+1, a built-in ability to substring would be very useful.
Or dynamically set run-name from job outputs

**Reply by chrisrosner** [reply]
Exactly the same use case here. This is the only thing I need it for. An alternative is if there were some additional special github values available like  {{ github.sha_8 }}

**Comment by abower-digimarc** [comment]
I'm also looking for a way to obtain the first 7 of the github_sha. I was first surprised that this value didn't exist in github contexts or environment variables, and secondly surprised that github yaml doesn't have any substring or length function. 

The only way to have that first 7, aka short sha, in multiple jobs, is to either recalculate it on every job, or add overhead to pass it between every job, both of which seem cumbersome.

**Reply by boorad** [reply]
like above, use the shell work-around
```yaml
    - name: Get short SHA
      id: get-short-sha
      run: |
        id=$(echo ${{github.sha}} | cut -c 1-7)
        echo "::set-output name=id::$id"
```

and then use it:

```yaml
      tags: ${{steps.get-short-sha.outputs.id}}
```

**Reply by dlech** [reply]
`::set-output` is deprecated, use `echo ${{github.sha}} | cut -c 1-7 >> $GITHUB_OUTPUT` instead

**Reply by dhirschfeld** [reply]
I'd like to use the short SHA in my `run_name` where I don't have access to either bash or any step outputs.

This use-case requires a substring function built into the GitHub Actions syntax.

**Reply by TWiStErRob** [reply]
@dhirschfeld I'm with you, in matrixes / inputs / workflow names / job names and probably many other places these solutions give us no recourse.

**Reply by dekimsey** [reply]
Agreed, basic string manipulation functions like substring(and frankly match/replace) would be immensely helpful to have. 

There was a recent question asking for this [here](https://github.com/orgs/community/discussions/48485) and while it was marked as completed it's a sub-optimal solution (as it required the OP to hard-code all environment names in their workflows).

**Comment by roryabraham** [comment]
I encountered this with concurrency keys. I wanted to do something like this:

```
concurrency:
  group: ${{ github.workflow }}-${{ split(github.ref_name, '-')[0] }}
  cancel-in-progress: true
```

Such that given semver versions like `1.0.0-1`, `1.0.0-2` and `1.0.1-0`, the concurrency key is generated only from the patch version `1.0.0` or `1.0.1`.

**Reply by roryabraham** [reply]
only workaround I could come up with is using a combination of:

- steps to generate the key (whatever scripting language you want)
- then conditionally use https://github.com/softprops/turnstyle for concurrency control. It's a great action and has served us well over the years but it ends up being more complex, and in fact doesn't support cancel-in-progress
- cancel-in-progress can be handled again by using a scripting language of your choice to use the GH API. I often reach for the gh cli

**Comment by dbsanfte** [comment]
So, is the lack of basic functions like `substring()` etc due to MS not wanting Github Actions to compete with their Azure Devops pipeline offering? Sure seems like it. Unreal that this still isn't available. Actions really feels half-finished.

**Comment by vi-shinde** [comment]
5 years now, still no ability to set run names with substring in expressions :(

**Comment by tommy-couzens** [comment]
:(
